packages/message-relayer/src/relay-tx.ts

```diff
@@ -0,0 +1,407 @@
+    throw new Error(`unable to find tx with hash: ${l2TransactionHash}`)
+  }
+
+  // Need to find the state batch for the given transaction. If no state batch has been published
+  // yet then we will not be able to generate a proof.
+  const batch = await getStateRootBatchByTransactionIndex(
+    l1RpcProvider,
+    l1StateCommitmentChainAddress,
+    l2Transaction.blockNumber - NUM_L2_GENESIS_BLOCKS
+  )
+  if (batch === null) {
+    throw new Error(
+      `unable to find state root batch for tx with hash: ${l2TransactionHash}`
+    )
+  }
+
+  const messages = await getMessagesByTransactionHash(
+    l2RpcProvider,
+    l2CrossDomainMessengerAddress,
+    l2TransactionHash
+  )
+
+  const messagePairs: CrossDomainMessagePair[] = []
+  for (const message of messages) {
+    // We need to calculate the specific storage slot that demonstrates that this message was
+    // actually included in the L2 chain. The following calculation is based on the fact that
+    // messages are stored in the following mapping on L2:
+    // https://github.com/ethereum-optimism/optimism/blob/c84d3450225306abbb39b4e7d6d82424341df2be/packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_L2ToL1MessagePasser.sol#L23
+    // You can read more about how Solidity storage slots are computed for mappings here:
+    // https://docs.soliditylang.org/en/v0.8.4/internals/layout_in_storage.html#mappings-and-dynamic-arrays
+    const messageSlot = ethers.utils.keccak256(
+      ethers.utils.keccak256(
+        encodeCrossDomainMessage(message) +
+          remove0x(l2CrossDomainMessengerAddress)
+      ) + '00'.repeat(32)
+    )
+
+    // We need a Merkle trie proof for the given storage slot. This allows us to prove to L1 that
+    // the message was actually sent on L2.
+    const stateTrieProof = await getStateTrieProof(
+      l2RpcProvider,
+      l2Transaction.blockNumber + NUM_L2_GENESIS_BLOCKS,
+      predeploys.OVM_L2ToL1MessagePasser,
+      messageSlot
+    )
+
+    // State roots are published in batches to L1 and correspond 1:1 to transactions. We compute a
+    // Merkle root for these state roots so that we only need to store the minimum amount of
+    // information on-chain. So we need to create a Merkle proof for the specific state root that
+    // corresponds to this transaction.
+    const txIndexInBatch =
+      l2Transaction.blockNumber - batch.header.prevTotalElements.toNumber()
+    const stateRootMerkleProof = getMerkleTreeProof(
+      batch.stateRoots,
+      txIndexInBatch
+    )
+
+    messagePairs.push({
+      message,
+      proof: {
+        stateRoot: batch.stateRoots[txIndexInBatch],
+        stateRootBatchHeader: batch.header,
+        stateRootProof: {
+          index: txIndexInBatch,
+          siblings: stateRootMerkleProof,
+        },
+        stateTrieWitness: stateTrieProof.accountProof,
+        storageTrieWitness: stateTrieProof.storageProof,
+      },
+    })
+  }
+
+  return messagePairs
```

Comment:
would pull this out into another function. I think there's a good chance that Etherscan will already have access to everything you queried above. and will just need this part :)

Reply:
Sounds good to me.

Reply:
Ok I looked into this more. It's a little annoying to break this logic out. Etherscan could probably input their own data here but that'll be harder than just using the function as given. Unless they're worried about performance I think it would be best to keep this function in one chunk for now.